content/docs/guides/custom-name-resolution.md

```diff
@@ -0,0 +1,104 @@
+
+Several languages support an interface to allow the user to define their own
+name resolver, so that you may define how to resolve any given name. Once
+registered, a name resolver with the _scheme_ `my-resolver` will be picked up
+when a target string begins with `my-resolver:`. For example, requests to
+`my-resolver:///my-service` would now use the `my-resolver` name resolver
+implementation.
+
+### Custom Name Resolvers
+
+You might consider using a custom name resolver whenever you would like to
+augment or replace DNS for service discovery. This interface has been used in
+the past to use [Apache Zookeeper](https://zookeeper.apache.org/) to look up
+service names.  It has also been used to directly interface with the Kubernetes
+API server for service lookup based on headless Service resources.
+
+One reason why it might be particularly useful to use a customer name resolver
+rather than standard DNS is that this interface is _reactive_. Within standard
+DNS, a client looks up the address for a particular service at the beginning of
+the connection and maintains its connection to that address for the lifetime
+of the connection. However, custom name resolvers may be watch-based. That is,
+they may maintain a connection to the name server for the duration of the
+connection and respond intelligently to backend failure as well as backend
+scale-ups and backend scale-downs.
+
+In addition, a custom name resolver may provide the client connection with a
+_service config_. A service config is essentially a JSON object that defines
+arbitrary configuration specifying how traffic should be routed to and load
+balanced across a particular service. At its most basic, this can be used to
+specify things like that a particular service should use the round robin load
+balancing policy vs. pick first. However, when a custom name resolver is used
+in conjunction with arbitrary service config and a [_custom load balancing
+policy_](custom-load-balancing.md), very complex traffic management systems
+such as xDS may be constructed.
+
+#### Life of a Target String
+
+While the exact interface for custom name resolvers differs from language to
+language, the general structure is the same. The client registers an
+implementation of a _name resolver provider_ to a process-global registry close
+to the start of the process. The name resolver provider will be called by the
+gRPC library with target strings intended for it. Given that target string, the
+name resolver provider will return an instance of a name resolver, which will
+interact with the client connection to direct the request according to the
+target string.
+
+```mermaid
+flowchart TB
+  Client -->|Request to my-resolver:///my-service| gRPC
+
+  gRPC -->|requests NameResolver| NameResolverProvider
+  NameResolverProvider -->|returns NameResolver| gRPC
+  gRPC -->|delegates resolution| NameResolver
+  NameResolver -->|connects| gRPC
```

Comment:
"connects"? Should it be "addresses"?

Reply:
This is actually something that differs between the Go interface and Java interface. Go provides the NameResolver with a ClientConnection object which the NameResolver is expected to mutate over time. Meanwhile, the Java interface periodically returns a set of resolved addresses. This was an attempt at papering over the interface difference.

Reply:
ClientConn is just a weird name. It isn't a connection or anything relating to subchannels. It is basically a view back into the Channel for callbacks. The important thing is you would call `UpdateState(State)` where `State` contains addresses when there is an update. That is essentially the same as Java's Listener2's `onResult(ResolutionResult)`. In both cases you pass a data object that holds addresses and service config info.